cmd/pullrequest-init/github.go

```diff
@@ -0,0 +1,227 @@
+	for _, c := range ic {
+		rawComment := filepath.Join(commentsPrefix, fmt.Sprintf("%d.json", c.GetID()))
+		h.Logger.Infof("Writing comment %d to file: %s", c.GetID(), rawComment)
+		if err := writeJSON(rawComment, c); err != nil {
+			return err
+		}
+
+		comment := &Comment{
+			Author: c.GetUser().GetLogin(),
+			Text:   c.GetBody(),
+			ID:     c.GetID(),
+
+			Raw: rawComment,
+		}
+		pr.Comments = append(pr.Comments, comment)
+	}
+
+	prPath := filepath.Join(path, prFile)
+	h.Logger.Infof("Writing pull request to file: %s", prPath)
+	return writeJSON(prPath, pr)
+}
+
+// readJSON reads an arbitrary JSON payload from path and decodes it into the
+// given interface.
+func readJSON(path string, i interface{}) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	return json.NewDecoder(f).Decode(i)
+}
+
+// Upload takes files stored on the filesystem and uploads new changes to
+// GitHub.
+func (h *GitHubHandler) Upload(ctx context.Context, path string) error {
+	h.Logger.Infof("Syncing path: %s to pr %d", path, h.prNum)
+
+	// TODO: Allow syncing from GitHub specific sources.
+
+	prPath := filepath.Join(path, prFile)
+	pr := new(PullRequest)
+	if err := readJSON(prPath, pr); err != nil {
+		return err
+	}
+
+	labelNames := make([]string, 0, len(pr.Labels))
+	for _, l := range pr.Labels {
+		labelNames = append(labelNames, l.Text)
+	}
+	h.Logger.Infof("Setting labels for PR %d to %v", h.prNum, labelNames)
+	if _, _, err := h.Issues.ReplaceLabelsForIssue(ctx, h.owner, h.repo, h.prNum, labelNames); err != nil {
+		return err
+	}
+
+	// Now sync comments.
+	desiredComments := map[int64]*Comment{}
+	for _, c := range pr.Comments {
+		desiredComments[c.ID] = c
+	}
+	h.Logger.Infof("Setting comments for PR %d to: %v", h.prNum, desiredComments)
+
+	existingComments, _, err := h.Issues.ListComments(ctx, h.owner, h.repo, h.prNum, nil)
+	if err != nil {
+		return err
+	}
+
+	for _, ec := range existingComments {
+		dc, ok := desiredComments[ec.GetID()]
+		if !ok {
+			// Delete
+			h.Logger.Infof("Deleting comment %d for PR %d", ec.GetID(), h.prNum)
+			if _, err := h.Issues.DeleteComment(ctx, h.owner, h.repo, ec.GetID()); err != nil {
+				return err
+			}
+		} else if dc.Text != ec.GetBody() {
+			//Update
+			newComment := github.IssueComment{
+				Body: github.String(dc.Text),
+				User: ec.User,
+			}
+			h.Logger.Infof("Updating comment %d for PR %d to %s", ec.GetID(), h.prNum, dc.Text)
+			if _, _, err := h.Issues.EditComment(ctx, h.owner, h.repo, ec.GetID(), &newComment); err != nil {
+				return err
+			}
+		}
+		// Delete to track new comments.
+		delete(desiredComments, ec.GetID())
+	}
+
+	for _, dc := range desiredComments {
+		newComment := github.IssueComment{
+			Body: github.String(dc.Text),
+		}
+		h.Logger.Infof("Creating comment %s for PR %d", dc.Text, h.prNum)
+		if _, _, err := h.Issues.CreateComment(ctx, h.owner, h.repo, h.prNum, &newComment); err != nil {
+			return err
```

Comment:
This loop will stop in the middle in case of error, and I don't see any easy way to make the upload idempotent (new comments do not have an ID), so I wonder if we should instead continue the loop, accumulate the errors and report back which comments failed instead of `err` only.

Reply:
Done!